packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m

```diff
@@ -478,52 +475,55 @@ - (void)picker:(PHPickerViewController *)picker
     [self sendCallResultWithSavedPathList:nil];
     return;
   }
-  dispatch_queue_t backgroundQueue =
-      dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0);
-  dispatch_async(backgroundQueue, ^{
-    NSNumber *maxWidth = self.callContext.maxSize.width;
-    NSNumber *maxHeight = self.callContext.maxSize.height;
-    NSNumber *imageQuality = self.callContext.imageQuality;
-    NSNumber *desiredImageQuality = [self getDesiredImageQuality:imageQuality];
-    NSOperationQueue *operationQueue = [NSOperationQueue new];
-    NSMutableArray *pathList = [self createNSMutableArrayWithSize:results.count];
-
-    for (int i = 0; i < results.count; i++) {
-      PHPickerResult *result = results[i];
-      FLTPHPickerSaveImageToPathOperation *operation = [[FLTPHPickerSaveImageToPathOperation alloc]
-               initWithResult:result
-                    maxHeight:maxHeight
-                     maxWidth:maxWidth
-          desiredImageQuality:desiredImageQuality
-                 fullMetadata:self.callContext.requestFullMetadata
-               savedPathBlock:^(NSString *savedPath) {
-                 pathList[i] = savedPath;
-               }];
-      [operationQueue addOperation:operation];
-    }
-    [operationQueue waitUntilAllOperationsAreFinished];
-    dispatch_async(dispatch_get_main_queue(), ^{
+  __block NSOperationQueue *saveQueue = [[NSOperationQueue alloc] init];
+  saveQueue.name = @"Flutter Save Image Queue";
+  saveQueue.qualityOfService = NSQualityOfServiceUserInitiated;
+
+  FLTImagePickerMethodCallContext *currentCallContext = self.callContext;
+  NSNumber *maxWidth = currentCallContext.maxSize.width;
+  NSNumber *maxHeight = currentCallContext.maxSize.height;
+  NSNumber *imageQuality = currentCallContext.imageQuality;
+  NSNumber *desiredImageQuality = [self getDesiredImageQuality:imageQuality];
+  BOOL requestFullMetadata = currentCallContext.requestFullMetadata;
+  NSMutableArray *pathList = [[NSMutableArray alloc] initWithCapacity:results.count];
```

Comment:
See my comment below: https://github.com/flutter/plugins/pull/6858/files#r1051290080

Reply:
Hmmm, maybe im missing something. I was saying if `pathList` is not populated (i.e. size=0, capacity=count), then will `pathList[0] = foo` throw index out of bounds?

Reply:
You're right, good catch. I'm going to make this a `NSPointerArray`.

Reply:
Actually just populating the array with `NSNull` as was done before is probably safer. A `NSPointerArray` won't have the out-of-bounds problem, since it "populates" up to `count` with `nil`, and will compact those out during `allObjects`.  So if there was a bug where all indices weren't set to `NSNull` then we would get wacky index values and the array length would be incorrect.

Reply:
Oh actually never mind it's the same logic either way, in either case the full result array is enumerated and populates the array or `NSPointerArray`, so there won't be stray `nil`s.

Reply:
Yes what I have here is correct, the first thing it's doing is populating the array with `NSNull`, in index order:
```objc
  NSMutableArray *pathList = [[NSMutableArray alloc] initWithCapacity:results.count];
...
  [results enumerateObjectsUsingBlock:^(PHPickerResult *result, NSUInteger index, BOOL *stop) {
    // NSNull means it hasn't saved yet.
    pathList[index] = [NSNull null];
...
```
This is exactly the same as the previous logic, it just replaced a for loop with fast enumeration over the results.
```objc
  NSMutableArray *mutableArray = [[NSMutableArray alloc] initWithCapacity:size];
  for (int i = 0; i < size; [mutableArray addObject:[NSNull null]], i++)
    ;
```

Reply:
Hmmm, the previous logic is using `addObject`, which increases the size though. iirc I think `pathList[index] = [NSNull null]` will simply crash.

Reply:
I just did an experiment. Very strange interesting/strange behavior of NSMutableArray: 

This worked: 

```
  NSMutableArray *arr = [NSMutableArray arrayWithCapacity:100];
  arr[0] = @"foo0";
  arr[1] = @"foo1";
  arr[2] = @"bar2";
  NSLog(@"print out: %@", arr);
```

This crashes: 

```
  NSMutableArray *arr = [NSMutableArray arrayWithCapacity:100];
  arr[0] = @"foo0";
  arr[2] = @"bar2";
  NSLog(@"print out: %@", arr);
```

This also crashes: 

```
  NSMutableArray *arr = [NSMutableArray arrayWithCapacity:100];
  arr[1] = @"foo1";
  arr[2] = @"bar2";
  NSLog(@"print out: %@", arr);
```

Looks like as long as you add from 0, one by one,  it auto expand the size. I wasn't able to find this behavior documented anywhere. However, [the doc of setObject API](https://developer.apple.com/documentation/foundation/nsmutablearray/1460093-setobject?language=objc) explicitly says it will crash if `index >= count`. So I recommend just do `addObject` instead of `setObject` API.

Reply:
The code is doing experiment 1 because `enumerateObjectsUsingBlock:` starts at 0 and goes through the whole array (unless `*stop = YES` is set).

https://developer.apple.com/documentation/foundation/nsarray/1415846-enumerateobjectsusingblock
> Executes a given block using each object in the array, starting with the first object and continuing through the array to the last object.

I can swap to `addObject:` though since it is incrementing by one every time.

Reply:
> I can swap to addObject: though since it is incrementing by one every time.

Yep, I think experiment 1 is a bug in `NSMutableArray` that's supposed to crash since it contradicts with the doc.